File: code/ruby/index1.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'xapian'
require 'json'
require_relative 'support'

### Start of example code.
def index_csv(data_path, db_path)
  db = Xapian::WritableDatabase.new(db_path, Xapian::DB_CREATE_OR_OPEN)
  term_generator = Xapian::TermGenerator.new
  term_generator.stemmer = Xapian::Stem.new('en')
  parse_csv_file(data_path).each do |row|
    doc = Xapian::Document.new
    term_generator.document = doc
    term_generator.index_text(row['TITLE'].to_s, 1, 'S')
    term_generator.index_text(row['DESCRIPTION'].to_s, 1, 'XD')
    term_generator.index_text(row['TITLE'].to_s)
    term_generator.increase_termpos
    term_generator.index_text(row['DESCRIPTION'].to_s)
    doc.data = row.to_h.to_json
    idterm = "Q#{row['id_NUMBER']}"
    doc.add_boolean_term(idterm)
    db.replace_document(idterm, doc)
  end
end
### End of example code.

abort "Usage #{__FILE__} DATAPATH DBPATH" if ARGV.length < 2

index_csv(ARGV[0], ARGV[1])

File: code/perl/index1.pl
#!/usr/bin/env perl

use strict;
use warnings;

BEGIN {
  eval {
    require Xapian;
    Xapian->import(':all');
    Xapian::search_xapian_compat();
  };
  if ($@) {
    require Search::Xapian;
    Search::Xapian->import(':all');
  }
}

use JSON::MaybeXS;
use FindBin qw($Bin);
use lib $Bin;
use Support;
use Data::Dumper;


my ($data_path, $db_path) = @ARGV;
die "Usage $0 DATAPATH DBPATH" unless $data_path && $db_path;

index_csv($data_path, $db_path);

### Start of example code.
sub index_csv {
    my ($data_path, $db_path) = @_;
    # Create or open the database we're going to be writing to.
    my $db = Search::Xapian::WritableDatabase->new($db_path, DB_CREATE_OR_OPEN);
    # Set up a TermGenerator that we'll use in indexing.
    my $term_generator = Search::Xapian::TermGenerator->new;
    $term_generator->set_stemmer(Search::Xapian::Stem->new('en'));
    foreach my $rec (Support::parse_csv($data_path)) {
        # print Dumper($rec);
        my $doc = Search::Xapian::Document->new;
        $term_generator->set_document($doc);
        # Index each field with a suitable prefix.
        $term_generator->index_text($rec->{TITLE}, 1, 'S');
        $term_generator->index_text($rec->{DESCRIPTION}, 1, 'XD');

        # Index fields without prefixes for general search.
        $term_generator->index_text($rec->{TITLE});
        $term_generator->increase_termpos();
        $term_generator->index_text($rec->{DESCRIPTION});

        # Store all the fields for display purposes.
        $doc->set_data(encode_json($rec));

        # We use the identifier to ensure each object ends up in the
        # database only once no matter how many times we run the
        # indexer.
        my $idterm = "Q" . $rec->{id_NUMBER};
        $doc->add_boolean_term($idterm);
        $db->replace_document_by_term($idterm, $doc);
    }
}
### End of example code.
